### PR TITLE
Capability to dynamically add Services to ServiceDataSource

### PR DESCRIPTION
**Pull request by gabber12**
## What?
Add `void add(Service service)` to ServiceDataSource

## Why?
RangerHub aims to be a single place to discover nodes for all services. In cases when the list of services are not known prehand, like usage of smart clients or dynamically created clients. Having capability to register service dynamically is also useful for services which cannot be known upfront.

Generally Smart clients have there own configuration management block. Service owners would have to do some wiring to make sure they merge smart client's configuration blocks with other config blocks to create rangerHub Service DataSource.
By adding these hooks, smart clients can dynamically add there service descriptors to Datasource dynamically

Addresses https://github.com/appform-io/ranger/issues/10

### CODE REVIEW

File: ranger-core/src/main/java/io/appform/ranger/core/finderhub/DynamicDataSource.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import lombok.NoArgsConstructor;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+@NoArgsConstructor
+public class DynamicDataSource implements ServiceDataSource {
+    private final Set<Service> serviceCollection = ConcurrentHashMap.newKeySet();
```
Comment (santanusinha) [inline ranger-core/src/main/java/io/appform/ranger/core/finderhub/DynamicDataSource.java:13]: Check if service is comparable

File: ranger-core/src/test/java/io/appform/ranger/core/finderhub/ServiceFinderHubTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                            .withNamespace(service.getNamespace())
+                            .withServiceName(service.getServiceName())
+                            .withDeserializer(new Deserializer<TestNodeData>() {
+                            })
+                            .build());
+
+    @Test
+    public void testDynamicServiceAddition() {
+        serviceFinderHub.start();
+        ServiceFinder<TestNodeData, MapBasedServiceRegistry<TestNodeData>> finder = serviceFinderHub.finder(new Service("NS", "PRE_REGISTERED_SERVICE"))
```
Comment (santanusinha) [inline ranger-core/src/test/java/io/appform/ranger/core/finderhub/ServiceFinderHubTest.java:38]: Use lombok stuff

File: ranger-core/src/test/java/io/appform/ranger/core/finderhub/ServiceFinderHubTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        val serviceFinderHub = new ServiceFinderHub<>(new StaticDataSource(new HashSet<>()), service -> new TestServiceFinderBuilder()
+                .withNamespace(service.getNamespace())
+                .withServiceName(service.getServiceName())
+                .withDeserializer(new Deserializer<TestNodeData>() {
+                })
+                .build());
+        serviceFinderHub.start();
+        try {
+            val future = serviceFinderHub.buildFinder(new Service("NS", "SERVICE_NAME"));
+            future.join();
```
Comment (santanusinha) [inline ranger-core/src/test/java/io/appform/ranger/core/finderhub/ServiceFinderHubTest.java:78]: fail if it comes here

**Review by santanusinha — CHANGES_REQUESTED** [review]

File: ranger-core/src/main/java/io/appform/ranger/core/finderhub/DynamicDataSource.java (hunk 1 above)
Comment (gabber12) [reply ranger-core/src/main/java/io/appform/ranger/core/finderhub/DynamicDataSource.java:13]: Service uses @Data annotation this include equalsandHashcode implementation

**Pull request merged**